File: src/interprocess/zeromq/julia/basic_multithread/multithread.jl
# Julia equivalent of src/interthread/basic_publisher_subscriber, showing the INTERTHREAD layer
# and pub/sub from a task rather than from Main.
#
# A publisher task produces NavigationReports on INTERTHREAD; a subscriber task consumes them,
# and publishes the deep ones straight out to INTERPROCESS. Those reach every application on the
# system that subscribed to the group (basic_julia_subscriber, in julia_multithread.launch), this
# one included: the publisher task subscribed to the same group and sees each one.
#
# Four things this example exists to show:
#
#   1. INTERTHREAD is pure Julia. Goby.jl carries these messages between tasks over Channels,
#      so they never enter the generated C++ and interthread groups are not declared in
#      interface.yml -- any string will do.
#   2. Because of that, an interthread message can be any Julia value, not only a protobuf
#      message. The alert Main receives below is a NamedTuple.
#   3. INTERPROCESS works from any task. Goby.jl hands the message to the task that owns the C++
#      application, the way a C++ thread's InterProcessForwarder does, so a task neither needs a
#      portal of its own nor has to route through Main.
#   4. Subscribing from a task works the same way: the subscription is registered on the task
#      that owns the application, and messages come back to the task that asked for them.
#
# Julia fixes its thread count at startup, and Goby.run() needs one thread per task module plus
# three (Main, the loop timer, and the C++ application). Two task modules therefore need five,
# which goby_add_julia_app()'s THREADS gives the launcher:
#
#   build/bin/basic_julia_multithread basic_julia_multithread.pb.cfg

include(joinpath(@__DIR__, "..", "goby_application_init.jl"))
init_cxxwrap("basic_julia_multithread")
include_protos(@__MODULE__, "goby3_examples")

# the groups and layer accessors generated from interface.yml, written beside this script by
# goby_add_julia_app()
include(joinpath(@__DIR__, "basic_julia_multithread_goby.jl"))
using .BasicJuliaMultithreadGoby: groups, interprocess

# interthread groups are Julia-side only, so they are named here rather than in interface.yml
const NAV_GROUP = "nav"
const DEPTH_ALERT_GROUP = "depth_alert"

# ---------------------------------------------------------------------------
# Publisher task: produces reports, and watches the ones that reach interprocess.
# ---------------------------------------------------------------------------
module PublisherTask

using Goby

function loop()
    nav = Main.goby3_examples.protobuf.NavigationReport(
        x = 95.0 + rand() * 20,
        y = 195.0 + rand() * 20,
        z = -305.0 + rand() * 10,
    )

    println("[publisher] Tx: z=$(round(nav.z, digits = 1))")
    Goby.publish(Main.app, Goby.INTERTHREAD, Main.NAV_GROUP, nav)
end

function published_nav(nav::Main.goby3_examples.protobuf.NavigationReport)
    println("[publisher] on the wire: z=$(round(nav.z, digits = 1))")
end

# subscriptions belong in start(), which Goby.run() calls on this task's own thread
function start()
    Goby.subscribe(Main.app, Main.interprocess(), Main.groups.julia_nav, published_nav)
end

# read by Goby.run() to drive loop() from the timer thread
goby_cfg = Dict(
    :loop_frequency => 1,   # Hz
    :loop_function => loop,
)

end # module PublisherTask

# ---------------------------------------------------------------------------
# Subscriber task: consumes reports, and sends the deep ones to the outside.
# ---------------------------------------------------------------------------
module SubscriberTask

using Goby

function incoming_nav(nav::Main.goby3_examples.protobuf.NavigationReport)
    println("[subscriber] Rx: z=$(round(nav.z, digits = 1))")

    if -nav.z > Main.alert_depth
        # straight out of this task: no portal here, and no hop through Main
        Goby.publish(Main.app, Main.interprocess(), Main.groups.julia_nav, nav)

        # an interthread message is any Julia value; this one never becomes protobuf
        Goby.publish(Main.app, Goby.INTERTHREAD, Main.DEPTH_ALERT_GROUP,
                     (depth = -nav.z, x = nav.x, y = nav.y))
    end
end

function start()
    Goby.subscribe(Main.app, Goby.INTERTHREAD, Main.NAV_GROUP, incoming_nav)
end

end # module SubscriberTask

# ---------------------------------------------------------------------------
# Main: reports the alerts the subscriber task raises.
# ---------------------------------------------------------------------------

function incoming_depth_alert(alert)
    println("[main] alert: $(round(alert.depth, digits = 1)) m at " *
            "x=$(round(alert.x, digits = 1)) y=$(round(alert.y, digits = 1))")
end

function start()
    Goby.subscribe(app, Goby.INTERTHREAD, DEPTH_ALERT_GROUP, incoming_depth_alert)
end

# how often the C++ application drains the channels the tasks publish to it through
goby_cfg = Dict(:cxx_channel_check_frequency => 10)

config_str = Goby.read_cli_cfg()
app = Goby.BasicJuliaMultithread(config_str)

# the application's own configuration, decoded into the Julia bindings generated from
# multithread_config.proto by goby_add_julia_protos()
pb_cfg = Goby.cfg(app, goby3_examples.config.BasicJuliaMultithreadConfig)
alert_depth = pb_cfg.alert_depth
println("Alerting below $(alert_depth) m")

Goby.run(app, Main, [PublisherTask, SubscriberTask])

File: src/interprocess/zeromq/python/basic_multithread/multithread.py
#!/usr/bin/env python3
"""Python version of src/interthread/basic_publisher_subscriber, showing the INTERTHREAD layer.

A publisher thread produces NavigationReports, a watcher thread consumes them and forwards each
one to the rest of the Goby system, and the application prints an alert when the vehicle is deep.

Three things this example exists to show:

  1. INTERTHREAD is pure Python. Goby carries these messages between threads without marshalling
     them, so they never enter the generated C++ and interthread groups are not declared in
     interface.yml -- any string will do. The application is a SingleThreadApplication: the C++
     side has no idea these threads exist.
  2. Because of that, an interthread message can be any Python object, not only a protobuf
     message. The alert below is a dataclass.
  3. INTERPROCESS belongs to the application's thread, but a thread publishes to it directly
     anyway: the publication is handed to the application behind the scenes, as a C++ thread's
     InterProcessForwarder hands its traffic to the portal.

Run it as any other Goby application:

    goby3_example_python_multithread --alert_depth 300
"""

import random
import sys
from dataclasses import dataclass

import goby

# generated from interface.yml by goby_add_python_app(); provides the application and thread base
# classes, and the groups declared in the interface file
from basic_python_multithread_goby import SingleThreadApplication, Thread, groups

# protoc --python_out of src/messages/nav.proto
from messages import nav_pb2

# interthread groups are Python-side only, so they are named here rather than in interface.yml
NAV_GROUP = "nav"
DEPTH_ALERT_GROUP = "depth_alert"


@dataclass
class DepthAlert:
    """An interthread message that is not a protobuf message, and never becomes one."""

    depth: float
    x: float
    y: float


class NavPublisher(Thread):
    """Produces reports on the interthread layer at its own rate."""

    def __init__(self):
        super().__init__(loop_frequency_hertz=1)

    def loop(self):
        nav = nav_pb2.NavigationReport(
            x=95 + random.randrange(20),
            y=195 + random.randrange(20),
            z=-305 + random.randrange(10),
        )

        print(f"[publisher] Tx: z={nav.z:.1f}")
        self.interthread().publish(NAV_GROUP, nav)


class DepthWatcher(Thread):
    """Consumes reports, forwards them outside the process, and alerts on deep ones."""

    def __init__(self):
        # no loop frequency: this thread is woken by its subscription rather than by a clock
        super().__init__()

        # subscribing in the constructor is what a C++ SimpleThread does, and works for the same
        # reason: the thread is constructed on the thread it runs on, so the callback lands here
        self.interthread().subscribe(NAV_GROUP, self.incoming_nav)

    def incoming_nav(self, nav: nav_pb2.NavigationReport):
        print(f"[watcher] Rx: z={nav.z:.1f}")

        # the interprocess portal lives on the application's thread; publishing to it from here
        # hands it over rather than touching the portal
        self.interprocess().publish(groups.nav, nav)

        if -nav.z > self.cfg.alert_depth:
            self.interthread().publish(
                DEPTH_ALERT_GROUP, DepthAlert(depth=-nav.z, x=nav.x, y=nav.y)
            )


class BasicMultithread(SingleThreadApplication):
    def __init__(self):
        # no loop frequency: launching a thread raises the rate the application is serviced at,
        # which is what picks up the alerts below
        super().__init__()

        print(f"Alerting below {self.cfg.alert_depth} m")

        self.interthread().subscribe(DEPTH_ALERT_GROUP, self.incoming_depth_alert)

        self.launch_thread(NavPublisher)
        self.launch_thread(DepthWatcher)

    # the annotation names the type this subscription takes, as it does for a protobuf message
    def incoming_depth_alert(self, alert: DepthAlert):
        print(f"[main] alert: {alert.depth:.1f} m at x={alert.x:.1f} y={alert.y:.1f}")


# reads command line parameters based on the BasicPythonMultithreadConfig definition, exactly as a
# C++ application does (try "--help" to see the parameters, or "--example_config" for the correct
# configuration file syntax). Edit "multithread_config.proto" to add parameters.
if __name__ == "__main__":
    sys.exit(goby.run(BasicMultithread))
